File: src/MixNetFx/Controls/ADNullableValueControls/ADNullableCheckBox.vb
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Drawing

Public Class ADNullableCheckBox
    Inherits ADNullableValueControlTemplate(Of Boolean)

    Protected Overrides Sub CreateControls()
        Me.EditableValueControl = New ADCheckBoxedValueForNullableValueControl
        Me.CaptionControl = New ADCaptionForNullableValueControl
    End Sub

    Sub New()
        MyBase.New()

        Dim locNullable As New ADDBNullable(Of Boolean)
        locNullable = Nothing

        ConsiderFixedSize = True
        myDontConditionForDisplay = True

        Me.Value = locNullable
    End Sub

    Protected Overrides Function GetInitialValueControlColor() As System.Drawing.Color
        Return SystemColors.Control
    End Function

    'Vorgegebene Breite der Controls
    'Aus ihnen ergibt sich das anfängliche Verhältnis für
    'Captionlänge und Wertebereichlänge
    Protected Overrides ReadOnly Property InitialCaptionControlLength() As Integer
        Get
            Return 400
        End Get
    End Property

    Protected Overrides ReadOnly Property InitialValueControlLength() As Integer
        Get
            Return 100
        End Get
    End Property

    <RefreshProperties(RefreshProperties.Repaint), _
    DefaultValue(True), _
    ADCategory("Verhalten", "Behaviour"), _
    ADDescription("Steuert, ob der Text im Datenerfassungsbereich mehr als eine Zeile umfassen darf.", _
               "Controls, if the text in the data input area can contain more than one line.")> _
    Public Property AutoHeight() As Boolean
        Get
            Return ConsiderFixedSize
        End Get
        Set(ByVal Value As Boolean)
            ConsiderFixedSize = Value
            UpdateLayout()
        End Set
    End Property

    Protected Overrides Function ToObjectForDisplaying(ByVal Value As IADDBNullableValue, ByVal ForSetValue As Boolean) As Object
        If Not Value.HasValue Then
            Return NullString
        Else
            Return Value.Value.ToString
        End If
    End Function

    Protected Overrides Function ToObjectForEditing(ByVal Value As IADDBNullableValue) As Object
        Return Value
    End Function

    Protected Overrides Function ToNullableValue(ByVal [Object] As Object) As ADDBNullable(Of Boolean)
        If [Object] Is Nothing Then
            Return Nothing
        End If

        Return CBool([Object])

    End Function
End Class

File: src/NetFxCSharp/Controls/ADNullableValueControls/ADNullableCheckBox.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace ActiveDev.Controls
{
    public class ADNullableCheckBox : ADNullableValueControlTemplate<bool>
    {
        protected override void CreateControls()
        {
            this.EditableValueControl = new ADCheckBoxedValueForNullableValueControl();
            this.CaptionControl = new ADCaptionForNullableValueControl();
        }

        public ADNullableCheckBox() : base()
        {
            ADDBNullable<bool> locNullable = new ADDBNullable<bool>();
            locNullable = default(ActiveDev.ADDBNullable<bool>);
            ConsiderFixedSize = true;
            myDontConditionForDisplay = true;
            this.Value = locNullable;
        }

        protected override System.Drawing.Color GetInitialValueControlColor()
        {
            return SystemColors.Control;
        }

        //Vorgegebene Breite der Controls
        //Aus ihnen ergibt sich das anfängliche Verhältnis für
        //Captionlänge und Wertebereichlänge
        protected override int InitialCaptionControlLength
        {
            get
            {
                return 400;
            }
        }

        protected override int InitialValueControlLength
        {
            get
            {
                return 100;
            }
        }

        [RefreshProperties(RefreshProperties.Repaint)]
        [DefaultValue(true)]
        [ADCategory("Verhalten", "Behaviour")]
        [ADDescription("Steuert, ob der Text im Datenerfassungsbereich mehr als eine Zeile umfassen darf.", "Controls, if the text in the data input area can contain more than one line.")]
        public bool AutoHeight
        {
            get
            {
                return ConsiderFixedSize;
            }

            set
            {
                bool Value = value;
                ConsiderFixedSize = Value;
                UpdateLayout();
            }
        }

        protected override object ToObjectForDisplaying(IADDBNullableValue Value, bool ForSetValue)
        {
            if (!(Value.HasValue))
            {
                return NullString;
            }
            else
            {
                return Value.Value.ToString();
            }
        }

        protected override object ToObjectForEditing(IADDBNullableValue Value)
        {
            return Value;
        }

        protected override ADDBNullable<bool> ToNullableValue(object Object)
        {
            if (Object == null)
            {
                return default(ActiveDev.ADDBNullable<bool>);
            }

            return System.Convert.ToBoolean(Object);
        }
    }
}